stan/math/prim/meta/holder.hpp

```diff
@@ -0,0 +1,274 @@
+  Holder(Holder<ArgType, Ptrs...>&&) = default;
+
+  Eigen::Index rows() const { return m_arg.rows(); }
+  Eigen::Index cols() const { return m_arg.cols(); }
+  Eigen::Index innerStride() const { return m_arg.innerStride(); }
+  Eigen::Index outerStride() const { return m_arg.outerStride(); }
+
+  auto* data() { return m_arg.data(); }
+
+  template <typename T, require_eigen_t<T>* = nullptr>
+  inline Holder<ArgType, Ptrs...>& operator=(const T& other) {
+    m_arg = other;
+    return *this;
+  }
+
+  inline Holder<ArgType, Ptrs...>& operator=(
+      const Holder<ArgType, Ptrs...>& other) {
+    m_arg = other;
+    return *this;
+  }
+
+  inline Holder<ArgType, Ptrs...>& operator=(Holder<ArgType, Ptrs...>&& other) {
+    m_arg = std::move(other.m_arg);
+    return *this;
+  }
+};
+
+}  // namespace math
+}  // namespace stan
+
+namespace Eigen {
+namespace internal {
+
+template <typename ArgType, typename... Ptrs>
+struct evaluator<stan::math::Holder<ArgType, Ptrs...>>
+    : evaluator_base<stan::math::Holder<ArgType, Ptrs...>> {
+  typedef stan::math::Holder<ArgType, Ptrs...> XprType;
+  typedef typename remove_all<typename nested_eval<ArgType, 1>::type>::type
+      ArgTypeNestedCleaned;
+  typedef typename XprType::CoeffReturnType CoeffReturnType;
+  typedef typename XprType::Scalar Scalar;
+  enum {
+    CoeffReadCost = evaluator<ArgTypeNestedCleaned>::CoeffReadCost,
+    Flags
+    = evaluator<ArgTypeNestedCleaned>::Flags
+      & (HereditaryBits | LinearAccessBit | PacketAccessBit | DirectAccessBit),
+    Alignment = evaluator<ArgTypeNestedCleaned>::Alignment,
+  };
+
+  evaluator<ArgTypeNestedCleaned> m_argImpl;
+
+  explicit evaluator(const XprType& xpr) : m_argImpl(xpr.m_arg) {}
+
+  EIGEN_STRONG_INLINE CoeffReturnType coeff(Index row, Index col) const {
+    return m_argImpl.coeff(row, col);
+  }
+  EIGEN_STRONG_INLINE CoeffReturnType coeff(Index index) const {
+    return m_argImpl.coeff(index);
+  }
+
+  EIGEN_STRONG_INLINE Scalar& coeffRef(Index row, Index col) {
+    return m_argImpl.coeffRef(row, col);
+  }
+  EIGEN_STRONG_INLINE Scalar& coeffRef(Index index) {
+    return m_argImpl.coeffRef(index);
+  }
+
+  template <int LoadMode, typename PacketType>
+  EIGEN_STRONG_INLINE PacketType packet(Index row, Index col) const {
+    return m_argImpl.template packet<LoadMode, PacketType>(row, col);
+  }
+  template <int LoadMode, typename PacketType>
+  EIGEN_STRONG_INLINE PacketType packet(Index index) const {
+    return m_argImpl.template packet<LoadMode, PacketType>(index);
+  }
+
+  template <int StoreMode, typename PacketType>
+  EIGEN_STRONG_INLINE void writePacket(Index row, Index col,
+                                       const PacketType& x) {
+    return m_argImpl.template writePacket(row, col, x);
+  }
+  template <int StoreMode, typename PacketType>
+  EIGEN_STRONG_INLINE void writePacket(Index index, const PacketType& x) {
+    return m_argImpl.template writePacket(index, x);
+  }
```

Comment:
I'm confused on why `packet` needs to pass along the templates but `writepacket` does not. Looking at some of the stuff based on `unary_evaluator` they pass writePacket at least the `StoreMode`

Reply:
It probably has some defaults. Those probably should be passed along.